feat(generate)!: initial generate ssp command

meganwolf0 opened the pull request:
## Description

Initial generate ssp command
* Adds `lula generate system-security-plan` with required profile and optional component path(s)
* (**Breaking**) Modifies the library ssp functionality to incorporate profile control resolution, instead of depending on controls defined by the component definitions (more information in below doc)
* Updates the `docs/oscal/system-security-plan.md` to reflect logic followed in the generate ssp command

Follow-on issues include:
- #810 - "Partial" compose is most likely the disaggregation of importing components to the component def and imported validations to the back-matter (follow-on discussion is potentially around continued import of validations to back-matter and utility)
- #624 - Injecting external, user defined, fields into the SSP (or other generated OSCAL models)

## Related Issue

Fixes #803

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [X] Test, docs, adr added or updated as needed
- [ ] (N/A) [Schema Updates](https://github.com/defenseunicorns/lula/blob/main/docs/community-and-contribution/schema-updates.md) applied
- [X] [Contributor Guide Steps](https://github.com/defenseunicorns/lula/blob/main/CONTRIBUTING.md) followed

mildwonkey commented on docs/oscal/system-security-plan.md:
```diff
@@ -1,25 +1,33 @@
 # System Security Plan
 
```
this is a really solid doc overall 👍🏻

meganwolf0 replied:
Thank you! @CloudBeard created this doc so props to him! I just updated some of it to reflect the new generation assumption 😄

mildwonkey commented on src/test/e2e/cmd/generate_ssp_test.go:
```diff
@@ -0,0 +1,145 @@
+		sspModel := complete.SystemSecurityPlan
+
+		require.NotNil(t, sspModel, "expected the SystemSecurityPlan model to be non-nil")
+		assert.Equal(t, 3, len(sspModel.ControlImplementation.ImplementedRequirements), "expected 3 controls")
```
I have no objection and certainly don't want anything you've written removed, but fwiw adding error messages to the asserts is rarely valuable IMO - if our error messages aren't clear enough to us you why the test failed, that's a sign to fix the error message in the test

meganwolf0 replied:
That's fair - I generally haven't been adding error messages and tbh I'm not really sure what came over me on this one that provided the inclination to adding them 😅 
That is a good point about these err messages potentially obfuscating _bad_ error output from the functions though 🤔

mildwonkey reviewed (commented):
I don't necessarily feel qualified to approve this (the oscal concepts are still so new to me), but I also don't have any objections and it all looks reasonable to me :) I left some comments but nothing that requires changes; feel free to dismiss 'em ❤️

brandtkeller commented on docs/oscal/system-security-plan.md:
```diff
@@ -1,25 +1,33 @@
 # System Security Plan
 
-A [System Security Plan](https://pages.nist.gov/OSCAL/resources/concepts/layer/implementation/ssp/) is an OSCAL-specific model to represent a system as a whole. In Lula, the `generate system-security-plan` command creates an `oscal-system-security-plan` object to explain the system as a whole by using the compliance data provided by the `component-definition`. The System Security Plan will detail ...
+A [System Security Plan](https://pages.nist.gov/OSCAL/resources/concepts/layer/implementation/ssp/) is an OSCAL-specific model to represent a system as a whole. In Lula, the `generate system-security-plan` command creates a representative `system-security-plan` object to describe the system as a whole. The only required input to the command is an OSCAL `profile` that defines the controls the syst ...
```
"system as a whole" feels like potentially repetitive language. We can probably drop that in favor of "the system".

brandtkeller commented on src/cmd/generate/system-security-plan.go:
```diff
@@ -0,0 +1,119 @@
+
+func GenerateSSPCommand() *cobra.Command {
+	var (
+		component  []string
```
[nit] Would plural `components` be more intuitive to the user?

meganwolf0 replied:
fair nit, I think you're right - will update

brandtkeller reviewed (commented):
Few minor comments thus far - have a little more tracing to do - otherwise looking great.

brandtkeller commented:
(few conflicts to resolve when you have a minute)

meganwolf0 commented on docs/cli-commands/lula_generate_system-security-plan.md:
```diff
@@ -0,0 +1,50 @@
+
+### Synopsis
+
+Generation of a System Security Plan OSCAL artifact from a source profile along with an optional list of component definitions.
```
This is a fair nitpicky comment - and I think I'd prefer the language `artifact` as it's a valid OSCAL SSP. Although I do see in generate component and profile we used the language of `template` (which is where I copy-pasta'd from)... Template now has a very specific denotation within Lula so I'm not really sure that's the right language to use and we should probably consolidate around another phrase for the `generate` commands. 

@brandtkeller might have some thoughts.

brandtkeller replied:
I would agree with your assessment. `artifact` is more aligned to the story Lula should tell in the long run. We've described it as a template in the past - as it's not a "complete" artifact - but it is still an artifact nonetheless.

meganwolf0 commented:
**Update**
* Resolved conflicts on main (with oscalTypes)
* Fixed some doc words
* Modified flag/var name for `component` -> `components` in generate SSP
* `template` -> `artifact` in short help for both SSP and Profile (Do we want to update generate component text as well OR is there another word we should give these things?)

Note - the `help.golden` files are kind of annoying since we have to re-generate the cli-docs + re-run the tests to generate the help docs when some command text words change... just throwing out the question of if those golden checks on `help` are valuable?

brandtkeller commented on src/cmd/generate/system-security-plan.go:
```diff
@@ -0,0 +1,119 @@
+				}
+
+				componentDefs = append(componentDefs, model.ComponentDefinition)
+				command += fmt.Sprintf(" --component %s", componentPath)
```
```suggestion
				command += fmt.Sprintf(" --components %s", componentPath)
```

meganwolf0 replied:
ahh a `component` hold out, nice catch!

brandtkeller reviewed (changes requested):
besides a typo I feel like this looks great and really sets us up well for AP generation.

mildwonkey reviewed (approved):
🎉
